src/v/cloud_storage/tests/cloud_storage_e2e_test.cc

```diff
@@ -583,3 +627,247 @@ FIXTURE_TEST(
+        .offset = model::offset(0),
+        .allow_compression = false,
+        .count = count,
+        .producer_id = pid.id,
+        .producer_epoch = pid.epoch,
+        .base_sequence = first_seq,
+        .is_transactional = is_transactional})};
+}
+
+void produce_until_segments(
+  int num_segments,
+  int tx_per_segment,
+  int batches_per_tx,
+  float abort_ratio,
+  cluster::partition* partition,
+  archival::ntp_archiver* archiver) {
+    cluster::rm_stm& stm = *partition->rm_stm();
+    ;
+    auto p = partition->get_ntp_config().ntp().tp.partition;
+
+    auto pid = model::producer_identity(42, 1);
+
+    for (int s_ix = 0; s_ix < num_segments; s_ix++) {
+        for (int tx = 0; tx < tx_per_segment; tx++) {
+            // start tx
+            auto res = stm.begin_tx(pid, model::tx_seq(0), 1s, p).get();
+            BOOST_REQUIRE(res.has_failure() == false);
+            auto tx_term = res.value();
+            vlog(e2e_test_log.info, "Tx term: {}", tx_term);
+
+            // produce shit
+            auto rrd = make_random_reader(pid, 0, batches_per_tx, true);
+            stm
+              .replicate(
+                rrd.id,
+                std::move(rrd.reader),
+                raft::replicate_options(raft::consistency_level::quorum_ack))
+              .get();
+
+            auto x = random_generators::get_real(0.0, 1.0);
+            if (x > abort_ratio) {
+                // commit tx
+                auto commit_res
+                  = stm.commit_tx(pid, model::tx_seq(0), 1s).get();
+                BOOST_REQUIRE(commit_res == cluster::tx_errc::none);
+            } else {
+                auto abort_res = stm.abort_tx(pid, model::tx_seq(0), 1s).get();
+                BOOST_REQUIRE_EQUAL(abort_res, cluster::tx_errc::none);
+            }
+        }
+
+        // Roll the segment
+        auto log = dynamic_cast<storage::disk_log_impl*>(
+          partition->log().get());
+        log->flush().get();
+        log->force_roll(ss::default_priority_class()).get();
+
+        BOOST_REQUIRE(archiver->sync_for_tests().get());
+        archiver->upload_next_candidates().get();
+    }
+
+    BOOST_REQUIRE_EQUAL(
+      cloud_storage::upload_result::success,
+      archiver->upload_manifest("test").get());
+    archiver->flush_manifest_clean_offset().get();
+
+    vlog(
+      e2e_test_log.info,
+      "Test manifest size is {} bytes",
+      partition->archival_meta_stm()->manifest().segments_metadata_bytes());
+}
+
+ss::future<bool> check_consume_from_beginning_no_check(
+  kafka::client::transport client,
+  const model::topic& topic_name,
+  ss::gate& gate) {
+    try {
+        kafka_consume_transport consumer(std::move(client));
+        co_await consumer.start();
+        int iters = 0;
+        while (iters == 0 || !gate.is_closed()) {
+            auto holder = gate.hold();
+            auto kvs = co_await consumer.consume_from_partition(
+              topic_name, model::partition_id(0), model::offset(0));
```

Comment:
Would be nice to check the returned offsets are correct (i.e. not part of an aborted tx).